File: AOC-2024-01A.hs
import Data.List (sort)

to_ints :: [String] -> [Int]
to_ints ws = map read ws

parse_lists ls = map to_ints (map words ls)

left ws = map head ws
right ws = map head $ map tail ws

main :: IO ()
main = do
  content <- getContents
  let lists = parse_lists (lines content)
  let lefts = sort $ left lists
  let rights = sort $ right lists

  let diffs = map abs $ zipWith (-) lefts rights
  let total = sum diffs
  print total

File: AOC-2024-01A.clj

(def input (slurp *in*))
(def lines (clojure.string/split-lines input))
(def pairs (map (fn [line] (clojure.string/split line #"\s+")) lines))

(def lefts (sort (map (fn [pair] (Integer/parseInt (nth pair 0))) pairs)))
(def rights (sort (map (fn [pair] (Integer/parseInt (nth pair 1))) pairs)))
(def diffs (map (fn [a b] (abs (- a b))) lefts rights))
(def total (reduce + diffs))

(println total)
